packages/scratch-vm/src/extensions/scratch3_spotify/index.ts

```diff
@@ -0,0 +1,696 @@
+ * - Reporter functions/blocks also take 0 or more arguments, but they must return a value (likely a `string` or `number`).
+ * - Hat functions/blocks also take 0 or more arguments, but they must return a boolean value.
+ *
+ * Feel free to delete these once you're ready to implement your own blocks.
+ *
+ * This type will be passed as the second generic argument to the Extension class that your specific extension 'extends'
+ * (see the class defintion below for more information on extending the Extension base class).
+ * @link https://www.typescriptlang.org/docs/handbook/2/functions.html Learn more about function types!
+ * @link https://www.typescriptlang.org/docs/handbook/2/objects.html Learn more about object types! (This is specifically a 'type alias')
+ * @link https://www.typescriptlang.org/docs/handbook/2/generics.html Learn more about generics!
+ */
+type Blocks = {
+    /*exampleCommand(exampleString: string, exampleNumber: number): void;
+  exampleReporter: () => number;
+  exampleHat(condition: boolean): boolean;*/
+    searchAndPlay(query: string): void;
+    searchAndPlayWait(query: string): void;
+    trackData(trackAttr: string): string;
+    stopMusic(): void;
+    // TODO uncomment when blocks are implemented
+    /* playNextBeat(): void;
+    playBeat(beatNum: number): void;
+    playBeatAndWait(beatNum: number): void;
+    currentBeat(): number;
+    everyBeat(): boolean;
+    everyBar(): boolean; */
+};
+
+/**
+ * @summary This is the class responsible for implementing the functionality of your blocks.
+ * @description You'll notice that this class `extends` (or 'inherits') from the base `Extension` class.
+ *
+ * Hover over `Extension` to get a more in depth explanation of the base class, and what it means to `extend it`.
+ */
+class Spotify extends Extension<Details, Blocks> {
+    /**
+     * @summary A field to demonstrate how Typescript Class fields work
+     * @link https://www.typescriptlang.org/docs/handbook/2/classes.html#fields
+     */
+    // Example field 1
+    //exampleField: number;
+    prevQuery: string;
+    currentArtistName: string;
+    currentTrackName: string;
+    currentAlbumName: string;
+
+    spotifyToken: AccessToken;
+
+    // FEEDBACK using JS library, not sure what type to put
+    player: any;
+    gain: any;
+    audioContext: any;
+
+    // TODO update all the "any"s with more descriptive types
+    beatPlayers: any[];
+    trackTimingData: any;
+    beatTimeouts: any[];
+    barTimeouts: any[];
+    trackTimeout: any;
+    trackStartTime: any;
+
+    releaseDur: number;
+    currentBeatPlayerIndex: number;
+    currentBeatNum: number;
+    beatFlag: boolean;
+    barFlag: boolean;
+    currentTrackDuration: number;
+    trackTempo: number;
+    numBeats: number;
+
+    async init(env: Environment) {
+        if (typeof Tone !== "undefined") {
+            this.prevQuery = "";
+            this.currentArtistName = "no artist";
```

Comment:
A very handy way to simplify these 'initializations' is using [initializers](https://www.typescriptlang.org/docs/handbook/2/classes.html#fields), which can also make the code more _obvious_ in that, by using _initializers_, it's clear we are _initializing_ those fields with a static value

So this:
```ts
class Spotify extends Extension<Details, Blocks> {
   prevQuery: string;
   constructor() {
      this. prevQuery = ""
   }
   ...
}
```
becomes this:
```ts
class Spotify extends Extension<Details, Blocks> {
   prevQuery: string = ""; // And you can actually even remove the string type, since it's clear this is a string from ow it's initialized
   constructor() {  }
   ...
}
```